// query-shape-detection/build/lib/constant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RDF_TRUE = exports.IRI_END_RDF_LIST = exports.IRI_REST_RDF_LIST = exports.IRI_FIRST_RDF_LIST = exports.TYPE_DEFINITION = exports.SHEX_ONE_OF = exports.SHEX_EACH_OF = exports.SHEX_VALUES = exports.SHEX_NON_LITERAL = exports.SHEX_BNODE = exports.SHEX_IRI = exports.SHEX_LITERAL = exports.SHEX_DATA_TYPE = exports.SHEX_MIN = exports.SHEX_MAX = exports.SHEX_NODE_KIND = exports.SHEX_VALUE_EXPR = exports.SHEX_CLOSED_SHAPE = exports.SHEX_EXPRESSIONS = exports.SHEX_SHAPE = exports.SHEX_EXPRESSION = exports.SHEX_SHAPE_EXPRESSION = exports.SHEX_PREDICATE = exports.DF = void 0;
const rdf_data_factory_1 = require("rdf-data-factory");
exports.DF = new rdf_data_factory_1.DataFactory();
exports.SHEX_PREDICATE = exports.DF.namedNode('http://www.w3.org/ns/shex#predicate');
exports.SHEX_SHAPE_EXPRESSION = exports.DF.namedNode('http://www.w3.org/ns/shex#shapeExpr');
exports.SHEX_EXPRESSION = exports.DF.namedNode('http://www.w3.org/ns/shex#expression');
exports.SHEX_SHAPE = exports.DF.namedNode('http://www.w3.org/ns/shex#Shape');
exports.SHEX_EXPRESSIONS = exports.DF.namedNode('http://www.w3.org/ns/shex#expressions');
exports.SHEX_CLOSED_SHAPE = exports.DF.namedNode('http://www.w3.org/ns/shex#closed');
exports.SHEX_VALUE_EXPR = exports.DF.namedNode('http://www.w3.org/ns/shex#valueExpr');
exports.SHEX_NODE_KIND = exports.DF.namedNode('http://www.w3.org/ns/shex#nodeKind');
exports.SHEX_MAX = exports.DF.namedNode('http://www.w3.org/ns/shex#max');
exports.SHEX_MIN = exports.DF.namedNode('http://www.w3.org/ns/shex#min');
exports.SHEX_DATA_TYPE = exports.DF.namedNode('http://www.w3.org/ns/shex#datatype');
exports.SHEX_LITERAL = exports.DF.namedNode('http://www.w3.org/ns/shex#literal');
exports.SHEX_IRI = exports.DF.namedNode('http://www.w3.org/ns/shex#iri');
exports.SHEX_BNODE = exports.DF.namedNode('http://www.w3.org/ns/shex#bnode');
exports.SHEX_NON_LITERAL = exports.DF.namedNode('http://www.w3.org/ns/shex#nonliteral');
exports.SHEX_VALUES = exports.DF.namedNode('http://www.w3.org/ns/shex#values');
exports.SHEX_EACH_OF = exports.DF.namedNode("http://www.w3.org/ns/shex#EachOf");
exports.SHEX_ONE_OF = exports.DF.namedNode("http://www.w3.org/ns/shex#OneOf");
exports.TYPE_DEFINITION = exports.DF.namedNode('http://www.w3.org/1999/02/22-rdf-syntax-ns#type');
exports.IRI_FIRST_RDF_LIST = exports.DF.namedNode('http://www.w3.org/1999/02/22-rdf-syntax-ns#first');
exports.IRI_REST_RDF_LIST = exports.DF.namedNode('http://www.w3.org/1999/02/22-rdf-syntax-ns#rest');
exports.IRI_END_RDF_LIST = exports.DF.namedNode('http://www.w3.org/1999/02/22-rdf-syntax-ns#nil');
exports.RDF_TRUE = exports.DF.literal('true', exports.DF.namedNode('http://www.w3.org/2001/XMLSchema#boolean'));
