// @comunica/actor-rdf-join-inner-hash/lib/ActorRdfJoinHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorRdfJoinHash = void 0;
const bus_rdf_join_1 = require("@comunica/bus-rdf-join");
const asyncjoin_1 = require("asyncjoin");
/**
 * A comunica Hash RDF Join Actor.
 */
class ActorRdfJoinHash extends bus_rdf_join_1.ActorRdfJoin {
    constructor(args) {
        super(args, {
            logicalType: 'inner',
            physicalName: 'hash',
            limitEntries: 2,
            requiresVariableOverlap: true,
        });
    }
    async getOutput(action) {
        let metadatas = await bus_rdf_join_1.ActorRdfJoin.getMetadatas(action.entries);
        // Ensure the left build stream is the smallest
        // TODO: in the next major version, use ActorRdfJoin.sortJoinEntries, which requires mediatorJoinEntriesSort
        if (metadatas[1].cardinality.value < metadatas[0].cardinality.value) {
            metadatas = [metadatas[1], metadatas[0]];
            action = { ...action, entries: [action.entries[1], action.entries[0]] };
        }
        const variables = bus_rdf_join_1.ActorRdfJoin.overlappingVariables(metadatas);
        const join = new asyncjoin_1.HashJoin(action.entries[0].output.bindingsStream, action.entries[1].output.bindingsStream, entry => bus_rdf_join_1.ActorRdfJoin.hash(entry, variables), bus_rdf_join_1.ActorRdfJoin.joinBindings);
        return {
            result: {
                type: 'bindings',
                bindingsStream: join,
                metadata: async () => await this.constructResultMetadata(action.entries, metadatas, action.context),
            },
        };
    }
    async getJoinCoefficients(action, metadatas) {
        // Ensure the left build stream is the smallest
        if (metadatas[1].cardinality.value < metadatas[0].cardinality.value) {
            metadatas = [metadatas[1], metadatas[0]];
        }
        const requestInitialTimes = bus_rdf_join_1.ActorRdfJoin.getRequestInitialTimes(metadatas);
        const requestItemTimes = bus_rdf_join_1.ActorRdfJoin.getRequestItemTimes(metadatas);
        return {
            iterations: metadatas[0].cardinality.value + metadatas[1].cardinality.value,
            persistedItems: metadatas[0].cardinality.value,
            blockingItems: metadatas[0].cardinality.value,
            requestTime: requestInitialTimes[0] + metadatas[0].cardinality.value * requestItemTimes[0] +
                requestInitialTimes[1] + metadatas[1].cardinality.value * requestItemTimes[1],
        };
    }
}
exports.ActorRdfJoinHash = ActorRdfJoinHash;
