// @comunica/actor-query-operation-distinct-hash/lib/ActorQueryOperationDistinctHash.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActorQueryOperationDistinctHash = void 0;
const bus_query_operation_1 = require("@comunica/bus-query-operation");
/**
 * A comunica Distinct Hash Query Operation Actor.
 */
class ActorQueryOperationDistinctHash extends bus_query_operation_1.ActorQueryOperationTypedMediated {
    constructor(args) {
        super(args, 'distinct');
    }
    async testOperation(_operation, _context) {
        return true;
    }
    async runOperation(operation, context) {
        const output = await this.mediatorQueryOperation.mediate({ operation: operation.input, context });
        // TODO: In next/major, remove undefined check when mediatorHashQuads is made required
        if (output.type === 'quads' && this.mediatorHashQuads !== undefined) {
            const outputQuads = bus_query_operation_1.ActorQueryOperation.getSafeQuads(output);
            const quadStream = outputQuads.quadStream.filter(await this.newHashFilterQuads(context));
            return {
                type: 'quads',
                quadStream,
                metadata: outputQuads.metadata,
            };
        }
        const outputBindings = bus_query_operation_1.ActorQueryOperation.getSafeBindings(output);
        const bindingsStream = outputBindings.bindingsStream.filter(await this.newHashFilter(context));
        return {
            type: 'bindings',
            bindingsStream,
            metadata: outputBindings.metadata,
        };
    }
    /**
     * Create a new distinct filter function.
     * This will maintain an internal hash datastructure so that every bindings object only returns true once.
     * @param context The action context.
     * @return {(bindings: Bindings) => boolean} A distinct filter for bindings.
     */
    async newHashFilter(context) {
        const { hashFunction } = await this.mediatorHashBindings.mediate({ allowHashCollisions: true, context });
        const hashes = {};
        return (bindings) => {
            const hash = hashFunction(bindings);
            return !(hash in hashes) && (hashes[hash] = true);
        };
    }
    /**
     * Create a new distinct filter function to hash quads.
     * This will maintain an internal hash datastructure so that every quad object only returns true once.
     * @param context The action context.
     * @return {(quad: RDF.Quad) => boolean} A distinct filter for quads.
     */
    async newHashFilterQuads(context) {
        // TODO: In next/major, this check can be removed when mediatorHashQuads is made required
        if (this.mediatorHashQuads === undefined) {
            return _quad => true;
        }
        const { hashFunction } = await this.mediatorHashQuads.mediate({ allowHashCollisions: true, context });
        const hashes = {};
        return (quad) => {
            const hash = hashFunction(quad);
            return !(hash in hashes) && (hashes[hash] = true);
        };
    }
}
exports.ActorQueryOperationDistinctHash = ActorQueryOperationDistinctHash;
